File: pydbn/interpreter/structures/frame.py

class DBNFrame(object):
    
    def __init__(self, parent=None, base_line_no=-1, return_pointer=-1, depth=0):
        self.base_line_no = base_line_no
        self.parent = parent
        self.env = {}

        self.stack = []
        self.return_pointer = return_pointer

        self.depth = depth

    def lookup_variable(self, key, default=None):
        """
        searches this first, then parents
        """
        try:
            return self.env[key]
        except KeyError:
            if self.parent is None:
                return default
            else:
                return self.parent.lookup_variable(key, default)
    
    def bind_variable(self, key, value):
        self.env[key] = value

    def bind_variables(self, **dct):
        self.env.update(dct)

    def __str__(self):
        return "[Frame|rp:%d|stack:%s|env:%s|parent:%s]" % (
            self.return_pointer,
            str(self.stack),
            str(self.env),
            str(self.parent),
        )

File: dbn.js/lib/structures/frame.js
define(function (require, exports, module) {
  "use strict";

  /**
   * DBNFrame
   */
  var DBNFrame = module.exports = function (options) {
    options = options || {};

    this.parent = options.parent || null;
    this.baseLineNo = options.baseLineNo == null ? -1 : options.baseLineNo;
    this.returnPointer = options.returnPointer == null ? -1 : options.returnPointer;
    this.depth = options.depth == null ? 0 : options.depth;

    this.env = {};
    this.stack = [];
  };
  
  (function () {
    this.lookupVariable = function (key, fallback) {
      if (typeof fallback === "undefined") {
        fallback = null;
      }

      if (this.env.hasOwnProperty(key)) {
        return this.env[key];
      } else {
        if (this.parent === null) {
          return fallback;
        } else {
          return this.parent.lookupVariable(key, fallback);
        }
      }
    };

    this.bindVariable = function (keyOrObject, value) {
      if (typeof keyOrObject === "string") {
        // then we assume we have a key, value pair
        this.env[keyOrObject] = value;
      } else if (typeof keyOrObject === "object") {
        // then its a set of key, value pairs to update
        var keysToSet = Object.getOwnPropertyNames(keyOrObject)
          , key
          , i
          ;

        for (i = 0; i < keysToSet.length; i++) {
          key = keysToSet[i];
          this.env[key] = keyOrObject[key];
        }
      }
    };

  }).call(DBNFrame.prototype);

});